etna/transforms/encoders/mean_encoder.py

```diff
@@ -0,0 +1,269 @@
+                    global_means_category[segment].pop(np.NaN, None)
+        else:
+            axis = None
+            global_means = nanmean(df.loc[:, self.idx[:, "target"]], axis=axis)
+
+            segment_df = TSDataset.to_flatten(df)
+            global_means_category = (
+                segment_df[[self.in_column, "target"]].groupby(self.in_column, dropna=False).mean().to_dict()["target"]
+            )
+            if self.handle_missing is MissingMode.global_mean:
+                global_means_category.pop(np.NaN, None)
+
+        self.global_means = global_means
+        self.global_means_category = global_means_category
+        return self
+
+    def _transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Get encoded values for the segment.
+
+        Parameters
+        ----------
+        df:
+            dataframe with data to transform.
+
+        Returns
+        -------
+        :
+            result dataframe
+
+        Raises
+        ------
+        ValueError:
+            If transform isn't fitted.
+        NotImplementedError:
+            If there are segments that weren't present during training.
+        """
+        if self.global_means is None:
+            raise ValueError("The transform isn't fitted!")
+
+        segments = df.columns.get_level_values("segment").unique().tolist()
+        n_segments = len(segments)
+        if self.mode is EncoderMode.per_segment:
+            new_segments = set(segments) - self.global_means.keys()
+            if len(new_segments) > 0:
+                raise NotImplementedError(
+                    f"This transform can't process segments that weren't present on train data: {reprlib.repr(new_segments)}"
+                )
+        current_timestamps = set(df.index)
+        intersected_timestamps = self.timestamps.intersection(current_timestamps)
+        future_timestamps = current_timestamps - self.timestamps
+
+        intersected_df = TSDataset.to_flatten(df.loc[list(intersected_timestamps)])
+        intersected_df = intersected_df.sort_values(["timestamp", "segment"])
+        intersected_df[self.out_column] = np.NaN
+
+        future_df = TSDataset.to_flatten(df.loc[list(future_timestamps)])
+        future_df = future_df.sort_values(["timestamp", "segment"])
+        future_df[self.out_column] = np.NaN
+
+        if len(intersected_df) > 0:
+            if self.mode is EncoderMode.per_segment:
+                for segment in segments:
+                    segment_df = intersected_df[intersected_df.segment == segment]
+                    y = segment_df["target"]
+                    temp = y.groupby(segment_df[self.in_column].astype(str)).agg(["cumsum", "cumcount"])
+                    feature = (temp["cumsum"] - y + self.global_means[segment] * self.smoothing) / (
+                        temp["cumcount"] + self.smoothing
+                    )
+                    intersected_df.loc[segment_df.index, self.out_column] = feature
```

Comment:
How could `intersected_df` not have `segment_df.index` in the timestamps?

Reply:
Fixed